confmap/provider/fileprovider/filewatcher.go

```diff
@@ -0,0 +1,175 @@
+		filePath:     filePath,
+		pollInterval: pollInterval,
+	}
+}
+
+// Start starts the watcher. It's safe to call multiple times - it's a noop if the watcher is already
+// running. It is, however, not thread-safe.
+func (fw *pollingFileWatcher) Start() error {
+	var err error
+
+	if fw.running.Load() {
+		return nil
+	}
+
+	fw.fileInfo, err = os.Stat(fw.filePath)
+	if err != nil {
+		return err
+	}
+
+	fileContent, err := os.ReadFile(fw.filePath)
+	if err != nil {
+		return err
+	}
+
+	fw.fileFingerprint = calculateFingerprint(fileContent)
+
+	fw.closeCh = make(chan struct{})
+	fw.Events = make(chan *confmap.ChangeEvent)
+
+	fw.wg.Add(1)
+	go func() {
+		defer fw.wg.Done()
+		fw.startWatchLoop()
+	}()
+
+	fw.running.Store(true)
+
+	return nil
+}
+
+// startWatchLoop starts the watch loop, which checks for changes to the file every poll interval
+// and emits events if needed
+// returns after closeCh is closed or after the context is cancelled
+func (fw *pollingFileWatcher) startWatchLoop() {
+	ticker := time.NewTicker(fw.pollInterval)
+	for {
+		select {
+		case <-ticker.C:
+			changeEvent := fw.checkFileForChanges()
+			if changeEvent != nil {
+				fw.sendEvent(changeEvent)
+			}
+		case _, ok := <-fw.closeCh:
+			if !ok {
+				return
+			}
+		}
+
+	}
+}
+
+func (fw *pollingFileWatcher) sendEvent(event *confmap.ChangeEvent) {
+	select {
+	case fw.Events <- event:
+	case _, ok := <-fw.closeCh:
+		if !ok {
+			return
+		}
+	}
+}
+
+// checkFileForChanges checks if the watched file changed since the previous check and returns an appropriate
+// ChangeEvent if it did, and nil if no changes were detected
+func (fw *pollingFileWatcher) checkFileForChanges() *confmap.ChangeEvent {
+	// check fileinfo first, it's cheaper than reading the content
+	fileInfo, err := os.Stat(fw.filePath)
+	if err != nil {
+		return &confmap.ChangeEvent{Error: err}
+	}
+	if metadataEqual(fileInfo, fw.fileInfo) {
+		return nil
+	}
+	fw.fileInfo = fileInfo
+
+	fileContent, err := os.ReadFile(fw.filePath)
+	if err != nil {
+		return &confmap.ChangeEvent{Error: err}
+	}
+
+	// check fingerprints
+	currentFingerprint := calculateFingerprint(fileContent)
+	if bytes.Equal(fw.fileFingerprint, currentFingerprint) {
+		return nil
+	}
+	fw.fileFingerprint = currentFingerprint
+
+	return &confmap.ChangeEvent{}
+}
+
+// Close stops the watcher. It's thread-safe and can be safely called multiple times.
+func (fw *pollingFileWatcher) Close() {
+	if !fw.running.CompareAndSwap(true, false) { // this is here to make it safe to call Close multiple times
+		return
+	}
+	close(fw.closeCh)
+	fw.wg.Wait()
+	close(fw.Events)
+}
+
+// calculateFingerprint calculates the fingerprint used to check for file changes
+func calculateFingerprint(content []byte) []byte {
+	hash := sha256.New()
+	hash.Write(content)
+	return hash.Sum(nil)
+}
+
+func metadataEqual(first os.FileInfo, second os.FileInfo) bool {
+	return first.Size() == second.Size() && first.ModTime() == second.ModTime()
```

Comment:
nit: some file file systems might be mounted with the `norelatime` option, which can lead to funny and unlikely bugs.

Reply:
Will it? It was my understanding that `norelatime` only affects access time, whereas I'm checking modification time here.

Reply:
My mistake, you are right. this will only affect the access time.